### ISSUE TITLE
Implement proper JAX/Numba dispatch for IfElse

### ISSUE THREAD

**Issue by ricardoV94** [pymc-devs/pytensor#501]
### Description

IfElse is only lazy in the default backend because the function virtual machine handles it (via the "lazy" attribute"). In Numba/JAX it currently does nothing, because it receives all outputs pre-computed.

During compilation we could specialize IfElse into a LazyIfElse Op that contains two InnerGraphs (maybe as OpFromGraph?), one corresponding to each branch. These graphs should contain all variables that lead to the inputs of IfElse and are not used by any other output variable other than through the outputs of IfElse. This depends on which function is being compiled and can't be known ahead of time.

The current implementation of jax_funcify_IfElse:
https://github.com/pymc-devs/pytensor/blob/4235ccc3f4243c5179178a206c15d84c4cda2e79/pytensor/link/jax/dispatch/basic.py#L60-L70

Would instead look something like (pseudo-code):
```python
@jax_funcify.register(LazyIfElse)
def jax_funcify_LazyIfElse(op, **kwargs):
    true_fn = jax_funcify(op.true_op.fgraph)
    false_fn = jax_funcify(op.false_op.fgraph)

    def ifelse(cond, *args):
        res = jax.lax.cond(cond, true_fn, false_fn, *args)
        return res if n_outs > 1 else res[0]

    return ifelse
```

This could even provide a nicer dprint, by showing the two inner graphs. Right now it's not always obvious what operations are lazily computed or not.

**Comment by Dhruvanshu-Joshi** [comment]
Hi 
Here's my understanding of what I think should be done:
Create a ```LazyIfElse``` op inheriting from the ```IfElse``` Op wherein we have seperate subgraphs for true and false conditional output essentially in the ```self.true_fgraph``` and ```self.false_fgraph```. 
Next we'll have to create a ```jax_funcify_LazyIfElse``` as you suggested.

Here's what I think should be the next pseudo-code:
```
def specialize_ifelse_to_lazy_ifelse(fgraph):
    for node in fgraph.apply_nodes:
        if isinstance(node.op, IfElse):
            condition = node.inputs[0]
            true_inputs = node.inputs[1:1 + node.op.n_outs]
            false_inputs = node.inputs[1 + node.op.n_outs:]
            
            outputs = node.outputs

            true_fgraph = create_subgraph(true_inputs, outputs[:node.op.n_outs])
            false_fgraph = create_subgraph(false_inputs, outputs[node.op.n_outs:])

            lazy_ifelse_op = LazyIfElse(true_fgraph, false_fgraph, node.op.n_outs)

            new_node = lazy_ifelse_op.make_node(condition, *node.inputs[1:])
            
            # Replace IfElse node with LazyIfElse node
            fgraph.replace_all([(old, new) for old, new in zip(node.outputs, new_node.outputs)])
```
Is this correct?
Also I cannot seem to understand the compilation process. How would I make the call for this ```specialize_ifelse_to_lazy_ifelse``` during compilation of jax/numba functions only and identify the op?

**Comment by ricardoV94** [comment]
So the challenge is to find the possible largest distinct subgraphs for the true and false branches that have the same set of inputs. This is a fun exercise in graph partitioning.

Here are two examples to illustrate this:
```python
cond = pt.scalar("cond", dtype=bool)

x = pt.scalar("x")
y = pt.exp(x)
out1 = pt.sin(y)
out2 = pt.cos(y) * x
out = ifelse(cond, out1, out2)
```

In that case we probably want true_fn and false_fn to be a function from (x, y) to out1/out2

```python
def true_func(x, y):
  return pt.sin(y)

def false_func(x, y):
  return pt.cos(y) * x
```

So operations that are shared between the two branches should be computed outside and given as inputs. At the same time, we want ewch function to have the largest number of unique operations inside each branch, to make them truly lazy

**Comment by Dhruvanshu-Joshi** [comment]
Hi @ricardoV94 
I was playing around with this yesterday for a while. What I did would makes sense more if I open a PR for that. Can you lease review this dprint that I am getting so that we know that I am on the right track:
```
if{inplace} [id A] 1
 ├─ cond [id B]
 ├─ Composite{...}.0 [id C] 0
 │  └─ x [id D]
 └─ Composite{...}.1 [id C] 0
    └─ ···

Inner graphs:

Composite{...} [id C]
 ← sin [id E] 'o0'
    └─ exp [id F] 't1'
       └─ i0 [id G]
 ← mul [id H] 'o1'
    ├─ cos [id I]
    │  └─ exp [id F] 't1'
    │     └─ ···
    └─ i0 [id G]
    ```

**Comment by ricardoV94** [comment]
Sounds good, although composite is a distinct op with InnerGraphs restricted to Elemwise operations.

We want LazyIfElse itself two have two inner graphs, like Composite has one

**Comment by ricardoV94** [comment]
Discussion from #1510 highlights we also need this approach if we want vectorized lazy ifelses

**Comment by ricardoV94** [comment]
For context in the C-backend laziness is achieved by how the VM is built. It computes nodes from outputs to inputs. When it arrives at a new node in requests it's ancestors be computed recursively, and then computes the node itself once this is done. For the IfElse, it computes the condition first, and then requests only the outputs of the respective branch. This naturally achieves laziness without any need to pre-order / pre-cluster nodes like the cond approach would need. 

They take advantage of this to also allow you to request only a subset of the function outputs be computed. It can just be implemented the same way.

This won't work for JAX, but could work for Numba (we could implement a very similar numba VM as well).
